### PR TITLE
Muir et al type blinding

### PR DESCRIPTION
**Pull request by slosar**
Progress so far:
 * Basic functionality to TwoPoint stage
 * Config plumbing and cosmology works 
 * Some work towards getting firecrown to make predictions
 * Comments welcome.

### CODE REVIEW

**Comment by joezuntz** [comment]
I don't think the code will be happy using the same name for the input and output files. 
You probably need to call the output one "twopoint_data_blind' or something.

**Comment by slosar** [comment]
This is now a working, e.g. running version of muir type blinding. It uses firecrown to make predictions. It requires the `parser_mods` branch of Firecrown. It hasn't really been validated, but I inspected a few numbers and they seem to be changing my sensible amounts. 
The interface to firecrown is somewhat klunky, but not terrible. 
For me this was really an exercise in how these things fit together.

File: txpipe/blinding.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
+###
+#
+# TO BE DEPRECATED VERY SOON
+# Leaving here to be able to see what is going on
```
Comment (joezuntz) [inline txpipe/blinding.py]: Delete this if it's not used now.

**Comment by joezuntz** [comment]
will add some more notes on the code itself though.

File: txpipe/blinding.py (hunk 1 above)
Comment (slosar) [reply txpipe/blinding.py]: Bah, I thought i've deleted that. Should be gone now.

**Comment by slosar** [comment]
The reason why I went against TXBlinding were:
 * You cannot make blinding default, so every person forgetting TXBlinding will start analyzing unblinded data. To me this just sounds a too easy barrier to breach by mistake
 * In my understanding each stage has an input and an output. To have TXBlinding as a separate stage, you need to have unblinded files lying around, which again are easy to breach. You could name them very prominently though!

Basically, if you see a way of enforcing TXBlinding stage unless you really want to disable it, then it might actually be not too bad.

File: txpipe/twopoint.py [hunk 2]
```diff
@@ -202,9 +221,118 @@ def _read_nbin_from_config(self):
...
+                                   sources: ['{tracer1}', '{tracer2}']
+                                   sacc_data_type: {dtype}
+                   """.format(fcname=fcname, dtype=dtype,tracer1=tracer1,tracer2=tracer2)
+        ## now try to get predictions
+        pred={}
+        for name,pars in [('fid',fid_params),('ofs',offset_params)]:
+            print ("Calling firecrown : %s"%(name))
+
+            config, data = firecrown.parse(io.StringIO(fc_config),
+               settings={'two_point':{'sacc_data':sacc},'parameters':pars})
```
Comment (beckermr) [inline txpipe/twopoint.py]: So @slosar simply make the edits you'd like to config and data here in the code. And then feed those to firecrown. This will keep the code I don't like out of firecrown.

Comment (slosar) [reply txpipe/twopoint.py]: Ok, I can pass cosmological parameters as strings rather than floats (it is ugly in my view, but ok), but I cannot pass a sacc object as a string. The only way to do it is dump it to disc, which is to start with very inefficient at nersc (takes minutes to dump a 2Mb sacc file), but important it is just wrong to be interacting with a library via objects on disk.

Comment (beckermr) [reply txpipe/twopoint.py]: Right so sacc object is def on the firecrown side. I will push a PR.

**Comment by joezuntz** [comment]
On the call we agreed to move this into TXBlinding

**Comment by slosar** [comment]
Now with a separate stage and with new firecrown interface (which does look much nicer!).
One side effect is that after TXBlinding deletes the unblided file, ceci wants to regenerate it next time. Not sure what to do about that. @joezuntz

**Comment by slosar** [comment]
Now it doesn't delete the unblinded file, but instead it rewrites it with zero size one.

File: txpipe/blinding.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+                }
+        }
+
+
+        for k,v in bz.items():
+            fc_config ['parameters']['bias_%s'%k] = v
+            
+        srclist=[]
+        lenslist=[]
+        for key,tracer in sacc.tracers.items():
```
Comment (joezuntz) [inline txpipe/blinding.py:112]: What you're trying to do for this whole chunk is not clear on first reading - can you add a lot more comments explaining what you're doing and how?

**Comment by joezuntz** [comment]
thought of some more things - hold fire.

File: txpipe/blinding.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+import warnings
+
+class TXBlinding(PipelineStage):
+    """
+    Blinding the data vectors.
+
+    """
+    name='TXBlinding'
+    inputs = [
+        ('twopoint_data_UNBLINDED_RABID', SACCFile),
```
[the remark this answers is not in the document]
Comment (joezuntz) [reply txpipe/blinding.py:13]: Actually - we do need a suffix here, because otherwise the unblind file will be used for later stages.
Argh.  Actually, I think I need to think about this.

Also in the master branch we've change `twopoint_data` -> `twopoint_data_real`

[the remark this answers is not in the document]
Comment (joezuntz) [reply txpipe/blinding.py:13]: Okay, think the answer here is to change the suffix on the original file to `_raw` and then also add a null blinding stage that just copies `twopoint_data_real_raw` -> `twopoint_data_real` (and overwrites the raw file with an empty one, as is done now).

Then `raw` doesn't indicate that it definitely will be blinded later, but is neutral on the question.

**Comment by ellongley** [comment]
Hi all, @chihway and I are volunteering to help on this branch as we need to learn to run the blinding stage for the reanalysis work. Here is an attempt at addressing the latest comments - feedback appreciated! :)

**Comment by slosar** [comment]
hi @ellongley  -- great to see progress. I really honestly had intention to finish this and will eventually, but if you beat me to it, even better! feel free to hassle me!

File: txpipe/blinding.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ellongley) [reply txpipe/blinding.py]: Are these now the correct suffixes? Is there a way to copy the file using the TXPipe object or should it just be with shutil e.g.?

File: txpipe/blinding.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        sys.stdout.flush()
+
+        if self.rank==0:
+            unblinded_fname = self.get_input('twopoint_data_UNBLINDED_RABID')
+            sacc = sacc.Sacc.load_fits(unblinded_fname)
+            blinded_sacc = self.blind_muir(sacc)
+            print ("Writing a very small sacc file [on NERSC: 1200 baud teetu-teetu-shhhhh]")
+            blinded_sacc.save_fits(self.get_output('twopoint_data'), overwrite=True)
+            if self.config['delete_unblinded']:
+                print ("Replacing %s with empty..."%unblinded_fname)
```
[the remark this answers is not in the document]
Comment (ellongley) [reply txpipe/blinding.py]: I _think_ I caught all the prints.

File: txpipe/blinding.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        
+         - Load two point SACC file
+         - Blinding it 
+         - Output blinded data
+         - Optionally deletete unblinded data
+        """
+        import sacc, sys, os
+        sys.stdout.flush()
+
+        if self.rank==0:
```
[the remark this answers is not in the document]
Comment (ellongley) [reply txpipe/blinding.py]: I deleted the test, but if there is something to add to mark to just run in serial that can be done. :)

**Comment by joezuntz** [comment]
One last request - could you add a null blinding step to the examples?
